### ISSUE TITLE
Run  A3c  (keras-rl/examples/a3c_pendulum.py),why the reward is so low?

### ISSUE THREAD

**Issue by coldsheephot** [keras-rl/keras-rl#114]
@matthiasplappert 
# first I run the example: keras-rl/examples/a3c_pendulum.py

# the result is :

<pre>
Interval 97 (960000 steps performed)
10000/10000 [==============================] - 13s - reward: -7.2284    
50 episodes - episode_reward: -1445.681 [-1657.401, -942.273] - loss: 1088.136 - mean_absolute_error: 26.070 - mean_q: -25.020 - mean_action: -48.453 - mean_action_variance: 25.289

Interval 98 (970000 steps performed)
10000/10000 [==============================] - 13s - reward: -7.1936    
50 episodes - episode_reward: -1438.725 [-1656.752, -494.390] - loss: 311.472 - mean_absolute_error: 13.869 - mean_q: -21.510 - mean_action: -33.922 - mean_action_variance: 18.609

Interval 99 (980000 steps performed)
10000/10000 [==============================] - 13s - reward: -7.1206    
50 episodes - episode_reward: -1424.117 [-1658.842, -728.142] - loss: 195.114 - mean_absolute_error: 11.244 - mean_q: -20.485 - mean_action: -34.226 - mean_action_variance: 16.894

Interval 100 (990000 steps performed)
10000/10000 [==============================] - 13s - reward: -7.1618    
done, took 4410.084 seconds
Testing for 5 episodes ...
Episode 1: reward: -1645.013, steps: 200
Episode 2: reward: -1540.265, steps: 200
Episode 3: reward: -1541.286, steps: 200
Episode 4: reward: -1430.997, steps: 200
Episode 5: reward: -1479.657, steps: 200
</pre>

# However in the openai gym :
The precise equation for reward:
<pre>
-(theta^2 + 0.1*theta_dt^2 + 0.001*action^2)
</pre>
Theta is normalized between -pi and pi. Therefore, the lowest cost is -(pi^2 + 0.1*8^2 + 0.001*2^2) = -16.2736044, and the highest cost is 0. In essence, the goal is to remain at zero angle (vertical), with the least rotational velocity, and the least effort.

#  do you run the example?why the  result is that !

**Comment by M00NSH0T** [comment]
I know this is really old, but I suspect it's because the layers aren't shared between the actor and critic.  The original A3C paper by DeepMind states that all the layers should be shared between the actors and critics, and I don't think the experimental version here is doing that.

**Comment by basarane** [comment]
I can't find it right now but I remember reading something like "if the policy and value function is linear wrt some features, the convergence is guaranteed." Also in the paper they state that all layers are shared between policy and value network except the last layer (softmax for policy, linear for value)

**Comment by M00NSH0T** [comment]
right yeah, all shared except the last layer (since once is a softmax vector vs a single scalar value for the Q value)

**Comment by basarane** [comment]
Another issue with A3C implementation might be related with error clipping. In the original paper the gradient of the loss is clipped rather than the loss itself. However the following function in a3c.py in experimental branch clips the error itself. This causes the gradients become zero outside of the delta range. 

        def clipped_mse(y_true, y_pred):
            delta = K.clip(y_true - y_pred, self.delta_range[0], self.delta_range[1])
            return K.mean(K.square(delta), axis=-1)

The following function can be replaced by hubber loss as described in the following post:
https://medium.com/@karpathy/yes-you-should-understand-backprop-e2f06eab496b

**Comment by ryou90** [comment]
[https://github.com/keras-rl/keras-rl/blob/experimental/examples/a3c_pendulum.py](url)

**Comment by nskplta** [comment]
Thanks! Did you manage to make it run with Keras 2.0? Seems many updates over A3C.py, still could not success.
